### ISSUE TITLE
`rego_type_error` may misreport offending rules on rule declaration conflict

### ISSUE THREAD

**Issue by johanfylling** [open-policy-agent/opa#5869]
Rules are not allowed to be declared within the extent of another rule, unless the "child" rule is assigning a value to a key in the partial object constructed by the "parent". E.g.:

```rego
package ex

p[k] := v { k := input.key; v := input.value } # 1
p.q := v { v := input.value } # 2
p.q.r := v { v := input.value } # 3
```

In the above policy, rule `1` is a partial object; rule `2` is assigning a value to the key `q` in object `p` (valid); rule `3` is inside the extent of rule `2`, but is not assigning a value to one of it's keys (invalid). The reported error is: 

```
2 errors occurred:
policy.rego:3: rego_type_error: single-value rule data.ex.p[k] conflicts with [data.ex.p.q]
policy.rego:4: rego_type_error: single-value rule data.ex.p.q conflicts with [data.ex.p.q.r]
```

`data.ex.p.q.r` (rule `3`) is indeed in conflict with both `data.ex.p[k]` (rule `1`) and `data.ex.p.q` (rule `2`); but rule `1`  is erroneously reported to be in conflict with rule `2`.

**Comment by ashutosh-narkar** [comment]
>  but rule 1 is erroneously reported to be in conflict with rule 2.
 
What happens if `input.key` is `q` in rule 1.  That would not be considered a conflict I think.

**Comment by johanfylling** [comment]
@ashutosh-narkar , indeed. And since the value is the same for rule 1 and 2, we don't even get an error at eval-time. But the presence of rule 3 makes this a compile-time error. For some reason, rules 1 and 2 are then also reported as a violation; which they won't be if rule 3 is removed from the policy.
